### PR TITLE
stub cellGem some more

### PR DESCRIPTION
**Pull request by Megamouse**
makes Top Darts show the calibration screen.

I'm not sure if all of this is working properly, let me know what you think.

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 1]
```diff
@@ -707,9 +810,27 @@ s32 cellGemInit(vm::cptr<CellGemAttribute> attribute)
...
-s32 cellGemInvalidateCalibration()
+s32 cellGemInvalidateCalibration(s32 gem_num)
 {
-	UNIMPLEMENTED_FUNC(cellGem);
+	cellGem.todo("cellGemInvalidateCalibration(gem_num=%d)", gem_num);
+	const auto gem = fxm::get<gem_t>();
+
+	if (!gem)
+	{
+		return CELL_GEM_ERROR_ALREADY_INITIALIZED;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: looks like i confused the error here

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 2]
```diff
@@ -765,9 +898,31 @@ s32 cellGemPrepareVideoConvert(vm::cptr<CellGemVideoConvertAttribute> vc_attribu
...
-	UNIMPLEMENTED_FUNC(cellGem);
+	cellGem.todo("cellGemReset(gem_num=%d, ext_id=*0x%x, ext_info=%s)", gem_num, ext_id, ext_info);
+	auto gem = fxm::get<gem_t>();
+
+	if (!gem)
+	{
+		return CELL_GEM_ERROR_UNINITIALIZED;
+	}
+
+	if (!check_gem_num(gem_num))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: should i check for !ext_id and !ext_info ?

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 3]
```diff
@@ -814,21 +969,72 @@ s32 cellGemSetRumble(u32 gem_num, u8 rumble)
 	return CELL_OK;
 }
 
-s32 cellGemSetYaw()
+s32 cellGemSetYaw(u32 gem_num, vm::ptr<f32[4]> z_direction)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: not sure if this is correct

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 4]
```diff
@@ -637,11 +711,11 @@ s32 cellGemGetState(u32 gem_num, u32 flag, u64 time_parameter, vm::ptr<CellGemSt
 		map_to_ds3_input(gem_num, gem_state->pad.digitalbuttons, gem_state->pad.analog_T);
 		map_ext_to_ds3_input(gem_num, gem_state->ext);
 
-		gem_state->tracking_flags = CELL_GEM_TRACKING_FLAG_POSITION_TRACKED |
-									CELL_GEM_TRACKING_FLAG_VISIBLE;
+		gem_state->tracking_flags = CELL_GEM_TRACKING_FLAG_POSITION_TRACKED | CELL_GEM_TRACKING_FLAG_VISIBLE;
 		gem_state->timestamp = gem->timer.GetElapsedTimeInMicroSec();
 
 		gem_state->quat[3] = 1.0;
+		return CELL_OK;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: this is probably the reason for Top Darts to progress to the next screen xD

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 5]
```diff
@@ -423,15 +441,39 @@ s32 cellGemForceRGB(u32 gem_num, float r, float g, float b)
...
+	const auto gem = fxm::get<gem_t>();
+
+	if (!gem)
+	{
+		return CELL_GEM_ERROR_UNINITIALIZED;
+	}
+
+	for (size_t i = 0; i < 360; i++)
+	{
+		hues[i] = 0;
```
Comment (VelocityRa) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: We're stubbing so why not `hues[i] = true;`?

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 6]
```diff
@@ -456,15 +498,47 @@ s32 cellGemGetCameraState(vm::ptr<CellGemCameraState> camera_state)
...
+	}
+
+	if (!r || !g || !b)
+	{
+		return CELL_GEM_ERROR_INVALID_PARAMETER;
+	}
+
+	//if (0)
+	//{
+	//	*r = 128;
```
Comment (VelocityRa) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: May as well use 128 for the color channels (games might special-case (0,0,0)). I have it the same in my branch.

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 4 above)
Comment (VelocityRa) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: yup lol

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 7]
```diff
@@ -667,15 +741,44 @@ s32 cellGemGetStatusFlags(u32 gem_num, vm::ptr<u64> flags)
...
-s32 cellGemGetTrackerHue()
+s32 cellGemGetTrackerHue(u32 gem_num, vm::ptr<u32> hue)
 {
-	UNIMPLEMENTED_FUNC(cellGem);
+	cellGem.warning("cellGemGetTrackerHue(gem_num=%d, hue=*0x%x)", gem_num, hue);
+	const auto gem = fxm::get<gem_t>();
+
+	if (!gem)
+	{
+		return CELL_GEM_ERROR_ALREADY_INITIALIZED;
```
Comment (VelocityRa) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: `CELL_GEM_ERROR_UNINITIALIZED`

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 1 above)
Comment (VelocityRa) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: yea

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 2 above)
Comment (VelocityRa) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: libgem only checks for `ext_id` apparently image

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 8]
```diff
@@ -804,7 +959,7 @@ s32 cellGemSetRumble(u32 gem_num, u8 rumble)
 		return CELL_GEM_ERROR_UNINITIALIZED;
 	}
 
-	if (!check_gem_num(gem_num))
+	if (!check_gem_num(gem_num) || rumble > 255)
```
Comment (VelocityRa) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: `rumble` is a `u8` so this is impossible.
libgem doesn't check anything regarding `rumble`.

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 3 above)
Comment (VelocityRa) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: probably is

File: rpcs3/Emu/Cell/Modules/cellGem.h [hunk 9]
```diff
@@ -18,6 +18,7 @@ enum
 	CELL_GEM_ERROR_CONVERT_NOT_STARTED        = 0x80121809,
 	CELL_GEM_ERROR_WRITE_NOT_FINISHED         = 0x8012180A,
 	CELL_GEM_ERROR_NOT_A_HUE                  = 0x8012180B,
+	//CELL_GEM_ERROR_LIGHTING_NOT_CALIBRATED    =;
```
Comment (VelocityRa) [inline rpcs3/Emu/Cell/Modules/cellGem.h]: ?

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 8 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: upsi. i just thought more than 255 is baka so I added that one without thinking

File: rpcs3/Emu/Cell/Modules/cellGem.h (hunk 9 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: what value does cellGemGetEnvironmentLightingColor return without calibration

Comment (VelocityRa) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: No such value exists in libgem. Seems like it just returns `CELL_OK` (with rgb to (128, 128, 128)).

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.h]: hmm ok. thx

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 5 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: yes. maybe it should

File: rpcs3/Emu/Cell/Modules/cellGem.cpp (hunk 6 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: okay I'll remove the scope and uncomment it

File: rpcs3/Emu/Cell/Modules/cellGem.cpp [hunk 10]
```diff
@@ -423,15 +441,39 @@ s32 cellGemForceRGB(u32 gem_num, float r, float g, float b)
 	return CELL_OK;
 }
 
-s32 cellGemGetAccelerometerPositionInDevice()
+s32 cellGemGetAccelerometerPositionInDevice(u32 gem_num, vm::ptr<f32[4]> pos)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellGem.cpp]: should this be a vm::pptr ?

Comment (VelocityRa) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: Oh yeah, I think it should. Maybe there are overloads for such arrays in `vm::ptr` but better safe than sorry.

Comment (VelocityRa) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: Actually no, isn't this just a pointer to 4 floats? Maybe it's fine as is.
Just leave whatever and I'll verify it at some point.

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: it should be different from the other occurance

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellGem.cpp]: since one is a getter and one is a setter

**Pull request merged**